### PR TITLE
Init time stats + Posts and Pages block

### PR DESCRIPTION
**Pull request by planarvoid**
This PR is introduces changes to architecture so that we can easily create use case instances with a different StatsGranularity (without the need for code duplication). UseCases are injected in `StatsModule`. I'm not 100% sure this approach is the best. Another way might be creating children of the `BaseListUseCase` and injecting these. The DI approach is very simple. Adding a new use case is just a matter of adding it into the injected list of use cases.

I had to create `PostsAndPagesUseCase` as part of this PR because it's the first time stats block and I had to have at least one to test this.

### CODE REVIEW

**Review by malinajirka — COMMENTED** [review]
Thanks @planarvoid! I like the approach, it just took me a while to understand everything. It might be worth adding some comments.

I've left several comments - some are not related to this PR and some are just nitpicks. Basically none of them is blocking. Let me know what you think.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/BaseListUseCase.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+constructor(
+    private val bgDispatcher: CoroutineDispatcher,
+    private val mainDispatcher: CoroutineDispatcher,
+    private val useCases: List<BaseStatsUseCase<*, *>>,
+    private val getStatsTypes: suspend (() -> List<StatsTypes>)
+) {
+    private val liveData = combineMap(
+            useCases.associateBy { it.type }.mapValues { entry -> entry.value.liveData }
+    )
+    private val insights = MutableLiveData<List<StatsTypes>>()
```
[the remark this answers is not in the document]
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/BaseListUseCase.kt]: good point, I'll rename it to `statsTypes`

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/StatsListViewModel.kt [hunk 2]
```diff
@@ -4,19 +4,32 @@ import android.arch.lifecycle.LiveData
...
 import org.wordpress.android.R
+import org.wordpress.android.ui.stats.refresh.lists.StatsUiState.StatsListState.DONE
+import org.wordpress.android.util.map
 import org.wordpress.android.viewmodel.ScopedViewModel
 
-abstract class StatsListViewModel(defaultDispatcher: CoroutineDispatcher) : ScopedViewModel(defaultDispatcher) {
+abstract class StatsListViewModel(defaultDispatcher: CoroutineDispatcher, private val statsUseCase: BaseListUseCase) :
+        ScopedViewModel(defaultDispatcher) {
+    private val _data = statsUseCase.data.map { loadedData ->
+        StatsUiState(loadedData, DONE)
```
[the remark this answers is not in the document]
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/StatsListViewModel.kt]: True, but we don't really use it at the moment so I'll remove the whole class and just use the `List<StatsBlock>` instead

File: WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java [hunk 3]
```diff
@@ -289,6 +289,19 @@ public static void viewTagsAndCategoriesStats(Context context, SiteModel site) {
         context.startActivity(intent);
     }
 
+    public static void viewPostsAndPagesStats(Context context, SiteModel site, StatsTimeframe statsTimeframe) {
+        Intent intent = new Intent(context, StatsViewAllActivity.class);
+        intent.putExtra(StatsAbstractFragment.ARGS_VIEW_TYPE, StatsViewType.TOP_POSTS_AND_PAGES);
+        intent.putExtra(StatsAbstractFragment.ARGS_TIMEFRAME, statsTimeframe);
+        intent.putExtra(StatsAbstractFragment.ARGS_SELECTED_DATE, "");
+        intent.putExtra(StatsAbstractFragment.ARGS_IS_SINGLE_VIEW, true);
+        intent.putExtra(StatsActivity.ARG_LOCAL_TABLE_SITE_ID, site.getId());
```
[the remark this answers is not in the document]
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java]: That's true but it's also used within the `StatsViewAllActivity` to read the param. We'll remove all of this once we implement the new `ViewAll` activity.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/BaseListUseCase.kt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    val data: LiveData<List<StatsBlock>> = mergeNotNull(insights, liveData) { insights, map ->
+        insights.mapNotNull { map[it] }
```
[the remark this answers is not in the document]
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/BaseListUseCase.kt]: Hmm, that makes sense. In the development phase I used this approach to hide the blocks that were not yet implemented.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/BaseListUseCase.kt [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import org.wordpress.android.util.mergeNotNull
+
+class BaseListUseCase
+constructor(
+    private val bgDispatcher: CoroutineDispatcher,
+    private val mainDispatcher: CoroutineDispatcher,
+    private val useCases: List<BaseStatsUseCase<*, *>>,
+    private val getStatsTypes: suspend (() -> List<StatsTypes>)
+) {
+    private val liveData = combineMap(
```
[the remark this answers is not in the document]
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/BaseListUseCase.kt]: I'm changing the field to `blockListData` - it's basically a map of type to list of blocks.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/StatsFragment.kt [hunk 6]
```diff
@@ -113,7 +114,7 @@ class StatsFragment : DaggerFragment() {
 
 class StatsPagerAdapter(val context: Context, val fm: FragmentManager) : FragmentPagerAdapter(fm) {
     companion object {
-        val statsTypes = listOf(INSIGHTS, DAYS, WEEKS, MONTHS)
+        val statsTypes = listOf(INSIGHTS, DAYS, WEEKS, MONTHS, YEARS)
```
[the remark this answers is not in the document]
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/StatsFragment.kt]: Great point! I'm renaming this to `StatsSection`. StatsTypes are coming from FluxC so I'd rather change it in another PR (if it still makes sense after this change)

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/StatsListFragment.kt [hunk 7]
```diff
@@ -197,17 +213,30 @@ class StatsListFragment : DaggerFragment() {
...
 
     class ViewFollowersStats : NavigationTarget()
     class ViewCommentsStats : NavigationTarget()
     class ViewTagsAndCategoriesStats : NavigationTarget()
     class ViewPublicizeStats : NavigationTarget()
     data class ViewTag(val link: String) : NavigationTarget()
+    data class ViewPostsAndPages(val statsGranularity: StatsGranularity) : NavigationTarget()
+}
+
+fun StatsGranularity.toStatsTimeFrame(): StatsTimeframe {
```
[the remark this answers is not in the document]
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/StatsListFragment.kt]: I removed the `dwmy`, it wasn't very clear anyway. I'm using either the `StatsGranularity` or `granular` instead.

**Comment by planarvoid** [comment]
@malinajirka thanks for the review! I've implemented the changes and it should be ready for another check

**Comment by malinajirka** [comment]
@planarvoid I see only this (83eb2c2) commit. Have you pushed all the changes?

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/BaseListUseCase.kt [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+            useCases.associateBy { it.type }.mapValues { entry -> entry.value.liveData }
+    )
+    private val insights = MutableLiveData<List<StatsTypes>>()
+    val data: LiveData<List<StatsBlock>> = mergeNotNull(insights, liveData) { insights, map ->
+        insights.mapNotNull { map[it] }
+    }
+
+    val navigationTarget: LiveData<NavigationTarget> = mergeNotNull(useCases.map { it.navigationTarget })
+
+    suspend fun loadInsightItems(site: SiteModel) {
```
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/BaseListUseCase.kt]: We should probably rename it. Same goes for `refreshInsightItems`.

**Review by malinajirka — APPROVED** [review]
Thanks @planarvoid! I've left one last comment, but it LGTM.

**Pull request merged**